Deactivating, reactivating, and deleting a FusionAuth user in Python (#837)
FusionAuth/fusionauth-site: site/_posts/2019-10-08-deactivating-reactivating-deleting-user-fusionauth-python.md

description: This is a step-by-step guide to deactivate, reactivate, or delete a user within your applications using API calls from the FusionAuth Python client library.
author: Amy Martin
image: blogs/python-fusionauth-deactivate.png
tags: client-python tutorial tutorial-python
category: blog
updated_date: 2021-07-23
excerpt_separator: "<!--more-->"
---

Did you know that you can deactivate, reactivate, or delete a user within your applications using API calls from the FusionAuth Python client library? These tools are great for data conscious organizations, executing your organization's data handling compliance policies (_[Learn more about GDPR data compliance here](/learn/expert-advice/ciam/developers-guide-to-gdpr)_), and fast user management for developers using FusionAuth. This tutorial will give us a chance to explore more of the FusionAuth Python client library while learning about these three valuable API calls.

<!--more-->

**Most recently updated July 23, 2021.**

## What we'll cover

1. Introduction to FusionAuth user management in Python
1. Verifying server applications, libraries, and dependencies
1. Obtaining a test user account Id

> 1. How to deactivate a user

Reviewer: Did you add any sections? I think so (kickstart)

```zsh
pip install fusionauth-client
```

Change directories to your preferred test application directory, and let's get started!

## Obtaining a test user account Id

For this tutorial, we'll first need a test user Id that is registered with FusionAuth.

**IMPORTANT:** If your FusionAuth server only has one admin user listed, you should definitely use a separate test user account for this tutorial. Create a new one before you continue.

In a web browser, let's go to "Users" in the FusionAuth UI. Select a test user, and under Action, navigate to "Manage", which should look something like this.

{% include _image.liquid src="/assets/img/blogs/action-manage-button.png" alt="Manage user" class="img-fluid" figure=false %}

On the "Manage User" page in the FusionAuth UI, you'll see the "User Id" located directly under the user's email address. In this instance, it's `4e0f2af0-33a2-4001-8631-0f8225e65f1c`.

Copy it into a text editor and set it off to the side. We'll be using it in a moment and throughout the rest of the tutorial.

{% include _image.liquid src="/assets/img/blogs/terry-test-account.png" alt="Terry test user info" class="img-fluid" figure=false %}

> ## Setting up FusionAuth using a Kickstart file

Reviewer: I'd add a section here defining each of these terms. Something like this (but jazz it up if you want).

* deactivate, also known as a soft delete or account locking, lets you disable the ability for a user to login temporarily
* reactivate removes the lock and lets a user login
* delete, also known as a hard delete, removes all data for a user and can't be undone. Beware!

Reviewer: looks good now.

# Deactivate A User with deactivate_user
client_response = client.deactivate_user('YOUR-USER-ID')
if client_response.was_successful():
  print(client_response.success_response)
else:
  print(client_response.error_response)
```

Save the program to your server, make sure that `venv` is running, and execute the application `deactivate.py`. If successful, you'll see a response of `None`.

Now, let's head on over to a web browser, and browse to "Users" so that we can check out our user. You should notice that the account associated with the user Id we provided is now locked. It will look something like this:

{% include _image.liquid src="/assets/img/blogs/terry-example-locked.png" alt="Terry is now locked" class="img-fluid" figure=false %}

> 

Reviewer: ```suggestion
When a user is locked, they can no longer log in to any applications, but they are not removed from the system. This is useful if you want to re-enable them later (perhaps after they've paid their bill).

```

client = FusionAuthClient('YOUR-OWN-API-KEY', 'http://localhost:9011')
```

Now let's call our `reactivate_user` function. Use the same user Id as you did in the previous example.

Add the `reactivate_user` function to our application. Be sure to replace the `reactivate_user` client argument of `YOUR-USER-ID` with your own selected user Id.

```python
# Reactivate A User
client_response = client.reactivate_user('YOUR-USER-ID')
if client_response.was_successful():
  print(client_response.success_response)
else:
  print(client_response.error_response)
```

Save and execute the `reactivate.py` application on your server. Upon success, you'll get a client response that should look something like this, displaying the user's profile information:

> ```zsh

Reviewer: Isn't this python?

Reviewer: please change it to python.

Author: fixed in latest commit

}
```

Refresh the "Users" page in your browser and you'll now notice that the lock icon has disappeared beside our test user and the account has been reactivated.

{% include _image.liquid src="/assets/img/blogs/terry-example-unlocked.png" alt="Terry is now unlocked" class="img-fluid" figure=false %}

## How to delete a user with FusionAuth Python client library

> For our next trick, we will be deleting a test user from the server. If you filled out a lot of profile information on your regular test user, you might want to create an empty test user for this tutorial. Just be sure to grab their "User Id" from the "Manage User" section of FusionAuth.

Reviewer: Why would you delete a user vs deactivate them? Did that get covered above (might have missed it).

{% include _image.liquid src="/assets/img/blogs/delete-test-user.png" alt="A user to delete" class="img-fluid" figure=false %}

Pop open your Python editor, and create a new file called `delete.py`. Import the FusionAuth client library, your API key, and your server info.

```python
from fusionauth.fusionauth_client import FusionAuthClient

# You must supply your API key and URL here

> client = FusionAuthClient('YOUR-API', 'http://localhost:9011')

Reviewer: ```suggestion
client = FusionAuthClient('YOUR-API-KEY', 'http://localhost:9011')
```

```

Now, add this section to the application, again replacing the `YOUR-USER-ID` with your test user's Id:

```python
# Delete user for a given Id
client_response =client.delete_user('YOUR-USER-ID')
if client_response.was_successful():
  print(client_response.success_response)
else:
  print(client_response.error_response)
```

Save and execute your `delete.py` application. (**Caution:** Executing this action will actually delete the user.)

> If successful, you will get a response of `None` and the user will now be deleted from FusionAuth. You can test this again by refreshing the "Users" page in your browser.  

Reviewer: Hmmm. Is this a hard delete or a soft delete? (Hard delete actually removes them from the system, soft delete lets you restore later.)  Looking at the source, it looks like a hard delete: https://github.com/FusionAuth/fusionauth-python-client/blob/master/src/main/python/fusionauth/fusionauth_client.py

**What this usually means:** Do you have the FusionAuth-Client Library already installed? If so, did you execute the application outside of venv? Browse to the directory containing venv, and activate `venv`:

```zsh
source venv/bin/activate
```

Then, change directories to your `fusionapps` folder and try again:

```zsh
python test.py
```

### Additional error codes

> Didn't find your error here? Visit [https://fusionauth.i/docs/v1/tech/apis/users](/docs/v1/tech/apis/users) for more response codes, or visit [https://fusionauth.io/docs/v1/tech/troubleshooting](/docs/v1/tech/troubleshooting) for more troubleshooting suggestions. You can also ask a question on the [FusionAuth Forum](https://fusionauth.io/community/forum/) or open an issue on [Github](https://github.com/FusionAuth/fusionauth-issues/issues/new/choose). As always, if you have a support contract feel free to [contact FusionAuth support](mailto:support@fusionauth.io).

Reviewer: ```suggestion
Didn't find your error here? Visit [https://fusionauth.i/docs/v1/tech/apis/users](/docs/v1/tech/apis/users) for more response codes, or visit [https://fusionauth.io/docs/v1/tech/troubleshooting](/docs/v1/tech/troubleshooting) for more troubleshooting suggestions. You can also ask a question on the [FusionAuth Forum](https://fusionauth.io/community/forum/). As always, if you have a support contract feel free to [contact FusionAuth support](https://account.fusionauth.io/account/support).
```
